src/main/scala/objsets/TweetSet.scala

```diff
@@ -0,0 +1,224 @@
+  /**
+    * Returns a new `TweetSet` which contains all elements of this set, and the
+    * the new element `tweet` in case it does not already exist in this set.
+    *
+    * If `this.contains(tweet)`, the current set is returned.
+    */
+  def incl(tweet: Tweet): TweetSet
+
+  /**
+    * Returns a new `TweetSet` which excludes `tweet`.
+    */
+  def remove(tweet: Tweet): TweetSet
+
+  /**
+    * Tests if `tweet` exists in this `TweetSet`.
+    */
+  def contains(tweet: Tweet): Boolean
+
+  /**
+    * This method takes a function and applies it to every element in the set.
+    */
+  def foreach(f: Tweet => Unit): Unit
+}
+
+class Empty extends TweetSet {
+
+  def filterAcc(p: Tweet => Boolean, acc: TweetSet): TweetSet = acc
+
+  def union(that: TweetSet): TweetSet = that
+
+  def mostRetweeted: Tweet = throw new NoSuchElementException
+
+  def isEmpty: Boolean = true
+
+  /**
+    * The following methods are already implemented
+    */
+  def contains(tweet: Tweet): Boolean = false
+
+  def incl(tweet: Tweet): TweetSet = new NonEmpty(tweet, new Empty, new Empty)
+
+  def remove(tweet: Tweet): TweetSet = this
+
+  def foreach(f: Tweet => Unit): Unit = ()
+}
+
+class NonEmpty(elem: Tweet, left: TweetSet, right: TweetSet) extends TweetSet {
+
+  def filterAcc(p: Tweet => Boolean, acc: TweetSet): TweetSet =
+    if (p(elem)) left.filterAcc(p, right.filterAcc(p, acc.incl(elem)))
+    else left.filterAcc(p, right.filterAcc(p, acc))
+
+  def union(that: TweetSet): TweetSet =
+    (left.union(right.union(that))).incl(elem)
+
+  def isEmpty: Boolean = false
+
+  def mostRetweeted: Tweet =
+    if (!left.isEmpty && left.mostRetweeted.retweets > elem.retweets) {
+      if (right.isEmpty || left.mostRetweeted.retweets > right.mostRetweeted.retweets) left.mostRetweeted
+      else right.mostRetweeted
+    } else if (!right.isEmpty && right.mostRetweeted.retweets > elem.retweets) right.mostRetweeted
+    else elem
+
+  /**
+    * The following methods are already implemented
+    */
+  def contains(x: Tweet): Boolean =
+    if (x.text < elem.text) left.contains(x)
+    else if (elem.text < x.text) right.contains(x)
+    else true
+
+  def incl(x: Tweet): TweetSet =
+    if (x.text < elem.text) new NonEmpty(elem, left.incl(x), right)
+    else if (elem.text < x.text) new NonEmpty(elem, left, right.incl(x))
+    else this
+
+  def remove(tw: Tweet): TweetSet =
+    if (tw.text < elem.text) new NonEmpty(elem, left.remove(tw), right)
+    else if (elem.text < tw.text) new NonEmpty(elem, left, right.remove(tw))
+    else left.union(right)
+
+  def foreach(f: Tweet => Unit): Unit = {
+    f(elem)
+    left.foreach(f)
+    right.foreach(f)
+  }
+}
+
+trait TweetList {
+  def head: Tweet
+  def tail: TweetList
+  def isEmpty: Boolean
+  def foreach(f: Tweet => Unit): Unit =
+    if (!isEmpty) {
+      f(head)
+      tail.foreach(f)
+    }
+}
+
+object Nil extends TweetList {
+  def head    = throw new java.util.NoSuchElementException("head of EmptyList")
```

Comment:
There's no need to write `java.util`. It can be done only with `NoSuchElementException`.